File: 543-diameter-of-binary-tree/diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int max;
    private int helper(TreeNode curr) {
        if (curr == null){
            return 0;
        }
        int left = helper(curr.left);
        int right = helper(curr.right);
        max = Math.max(max, left+right);
        return 1+Math.max(left, right);
    }



    public int diameterOfBinaryTree(TreeNode root) {
        helper(root);
        return max;
    }

}

File: 543-diameter-of-binary-tree/diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.result = 0
        def helper(root):
            if not root:
                return 0
            left = helper(root.left)
            right = helper(root.right)
            self.result = max(self.result, left + right)
            return max(left, right) + 1
        helper(root)
        return self.result
            


        